src/code/z_sub_s.c

```diff
@@ -333,13 +341,189 @@ s32 SubS_CopyPointFromPathCheckBounds(Path* path, s32 pointIndex, Vec3f* dst) {
+        { 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0 },
+        { 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0 },
+        { 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0 },
+        { 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0 },
+        { 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0 },
+        { 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0 },
+        { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+        { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+    },
+    {
+        { 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0 },
+        { 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0 },
+        { 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0 },
+        { 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0 },
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 },
+        { 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0 },
+        { 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0 },
+        { 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0 },
+        { 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0 },
+    },
+};
+
+void SubS_FillShadowTex(s32 startCol, s32 startRow, u8* tex, s32 size) {
+    s32 i;
+    s32 j;
+    s32 start;
+
+    for (i = 0; i < 12; i++) {
+        start = ((startRow + i) * 64) + startCol - 390;
+        for (j = 0; j < 12; j++) {
+            if (sShadowMaps[size][i][j] != 0) {
+                if ((start + j >= 0) && (start + j < (s32)sizeof(u8[64][64]))) {
+                    tex[start + j] = 255;
+                }
+            }
+        }
+    }
+}
+
+void SubS_GenShadowTex(Vec3f bodyPartsPos[], Vec3f* worldPos, u8* tex, f32 tween, u8 bodyPartsNum, u8 sizes[],
+                       s8 parentBodyParts[]) {
+    Vec3f pos;
+    Vec3f startVec;
+    s32 i;
+    s32 parentBodyPart;
+    Vec3f* bodyPartPos;
+    s32 startCol;
+    s32 startRow;
+
+    for (i = 0; i < bodyPartsNum; i++) {
+        if (parentBodyParts[i] >= 0) {
+            parentBodyPart = parentBodyParts[i];
+            bodyPartPos = &bodyPartsPos[i];
+
+            pos.x = (bodyPartsPos[parentBodyPart].x - bodyPartPos->x) * tween + (bodyPartPos->x - worldPos->x);
+            pos.y = (bodyPartsPos[parentBodyPart].y - bodyPartPos->y) * tween + (bodyPartPos->y - worldPos->y);
+            pos.z = (bodyPartsPos[parentBodyPart].z - bodyPartPos->z) * tween + (bodyPartPos->z - worldPos->z);
+        } else {
+            bodyPartPos = &bodyPartsPos[i];
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sub_s/func_8013CF04.s")
+            pos.x = bodyPartPos->x - worldPos->x;
+            pos.y = bodyPartPos->y - worldPos->y;
+            pos.z = bodyPartPos->z - worldPos->z;
+        }
+
+        Matrix_MultiplyVector3fByState(&pos, &startVec);
+        startCol = 64.0f + startVec.x;
+        startRow = 64.0f - startVec.z;
+        SubS_FillShadowTex(startCol >> 1, startRow >> 1, tex, sizes[i]);
+    }
+}
+
+void SubS_DrawShadowTex(Actor* actor, GraphicsContext** gfxCtxPtr, u8* tex) {
+    s32 pad;
+    GraphicsContext* gfxCtx = *gfxCtxPtr;
```

Comment:
Random question (not saying it would be better), would this match with a `GameState` param instead?
```suggestion
void SubS_DrawShadowTex(Actor* actor, GameState* gameState u8* tex) {
    s32 pad;
    GraphicsContext* gfxCtx = gameState->gfxCtx;
```

Reply:
Oh, didn't realize`gfxCtx` was the first struct member of `GameState`.
Changing to `GameState*` does work, so does `GlobalContext*`, either of which would probably be preferable. Any preference on those two? I'd probably lean more towards `GlobalContext*`, but don't have a strong opinion either way.

Reply:
Went with `GlobalContext*`. If we would prefer `GameState*` just let me know.